src/Microsoft.AspNetCore.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs

```diff
@@ -307,32 +317,111 @@ private IEnumerable<string> GetAttributeNames(Block tagBlock)
+                return Enumerable.Empty<KeyValuePair<string, string>>();
             }
 
-            var attributeChildren = new List<SyntaxTreeNode>(childCount - 1);
-            for (var i = 1; i < childCount; i++)
-            {
-                attributeChildren.Add(tagBlock.Children[i]);
-            }
-            var attributeNames = new List<string>();
+            _htmlAttributeTracker.Clear();
+
+            var attributes = _htmlAttributeTracker;
 
-            foreach (var child in attributeChildren)
+            for (var i = 1; i < childCount; i++)
             {
+                var child = tagBlock.Children[i];
                 Span childSpan;
 
                 if (child.IsBlock)
                 {
-                    childSpan = ((Block)child).FindFirstDescendentSpan();
+                    var childBlock = (Block)child;
+
+                    if (childBlock.Type != BlockType.Markup)
+                    {
+                        // Anything other than markup blocks in the attribute area of tags mangles following attributes.
+                        // It's also not supported by TagHelpers, bail early to avoid creating bad attribute value pairs.
+                        break;
+                    }
+
+                    childSpan = childBlock.FindFirstDescendentSpan();
 
                     if (childSpan == null)
                     {
+                        _attributeValueBuilder.Append(InvalidAttributeValueMarker);
                         continue;
                     }
+
+                    var childOffset = 0;
+
+                    if (childSpan.Symbols.Count > 0)
+                    {
+                        var potentialQuote = childSpan.Symbols[childSpan.Symbols.Count - 1] as HtmlSymbol;
+                        if (potentialQuote != null &&
+                            (potentialQuote.Type == HtmlSymbolType.DoubleQuote ||
+                            potentialQuote.Type == HtmlSymbolType.SingleQuote))
+                        {
+                            childOffset = 1;
+                        }
+                    }
+
+                    for (var j = 1; j < childBlock.Children.Count - childOffset; j++)
+                    {
+                        var valueChild = childBlock.Children[j];
+                        if (valueChild.IsBlock)
+                        {
+                            _attributeValueBuilder.Append(InvalidAttributeValueMarker);
+                        }
+                        else
+                        {
+                            var valueChildSpan = (Span)valueChild;
+                            for (var k = 0; k < valueChildSpan.Symbols.Count; k++)
+                            {
+                                _attributeValueBuilder.Append(valueChildSpan.Symbols[k].Content);
+                            }
+                        }
+                    }
                 }
                 else
                 {
                     childSpan = child as Span;
```

Comment:
Actually, since this is soon dereferenced, this is important. NREs are much harder to debug that `InvalidCastException`s.